tiledb/sm/c_api/tiledb.h

```diff
@@ -3201,6 +3234,468 @@ TILEDB_EXPORT int32_t tiledb_array_encryption_type(
     const char* array_uri,
     tiledb_encryption_type_t* encryption_type);
 
+/* ****************************** */
+/*            SUBARRAY            */
+/* ****************************** */
+
+/**
+ * Creates a subarray object of an open array.
+ *
+ * **Example:**
+ *
+ * @code{.c}
+ * tiledb_array_t* array;
+ * tiledb_array_alloc(ctx, "file:///my_array", &array);
+ * tiledb_array_open(ctx, array, TILEDB_READ);
+ * tiledb_subarray_t* subarray;
+ * tiledb_subarray_alloc(ctx, array, TILEDB_UNORDERED, &subarray);
+ * @endcode
+ *
+ * @param ctx The TileDB context.
+ * @param array The array that the subarray corresponds to.
+ * @param layout The layout of either the results if the subarray is
+ *     used for reading, or the layout of the values that the user will provide
+ *     for writing.
+ * @param subarray The subarray object to be created.
+ * @return `TILEDB_OK` for success and `TILEDB_OOM` or `TILEDB_ERR` for error.
+ */
+TILEDB_EXPORT int32_t tiledb_subarray_alloc(
+    tiledb_ctx_t* ctx,
+    tiledb_array_t* array,
+    tiledb_layout_t layout,
+    tiledb_subarray_t** subarray);
+
+/**
+ * Deletes a subarray object.
+ *
+ * @param subarray The subarray object to be deleted.
+ */
+TILEDB_EXPORT void tiledb_subarray_free(tiledb_subarray_t** subarray);
+
+/**
+ * Retrieves the subarray layout.
+ *
+ * **Example:**
+ *
+ * @code{.c}
+ * tiledb_layout_t layout;
+ * tiledb_subarray_get_layout(ctx, subarray, &layout);
+ * @endcode
+ *
+ * @param ctx The TileDB context.
+ * @param subarray The subarray.
+ * @param layout The layout to be retrieved.
+ * @return `TILEDB_OK` for success and `TILEDB_ERR` for error.
+ */
+TILEDB_EXPORT int32_t tiledb_subarray_get_layout(
+    tiledb_ctx_t* ctx,
+    const tiledb_subarray_t* subarray,
+    tiledb_layout_t* layout);
+
+/**
+ * Retrieves the subarray type.
```

Comment:
I know that these functions were primarily for testing, but to me it is confusing that a subarray object has a datatype and a domain, since they have to be the same as what's in the array schema. If you want to keep these though I suggest this function be renamed `tiledb_subarray_get_datatype`.